packages/ui/feature-pieces/src/lib/services/piece.service.ts

```diff
@@ -0,0 +1,231 @@
+import { DropdownState, PiecePropertyMap } from "@activepieces/pieces-framework";
+import { Action, ActionType, AddPieceRequestBody, ListPiecesRequestQuery, PieceCategory, PieceOptionRequest, PieceScope, SuggestionType, Trigger, TriggerType, isNil, spreadIfDefined } from "@activepieces/shared";
+import { AuthenticationService, FlowItemDetails, PieceMetadataModel, PieceMetadataModelSummary, environment } from "@activepieces/ui/common";
+import { HttpClient } from "@angular/common/http";
+import { Injectable } from "@angular/core";
+import { Observable, Subject, map, of, shareReplay } from "rxjs";
+
+
+@Injectable({
+    providedIn: 'root',
+})
+export class PieceMetadataService {
+
+    private CODE_ITEM_DETAILS = {
+        type: ActionType.CODE,
+        name: 'Code',
+        description: 'Powerful nodejs & typescript code with npm',
+        logoUrl: '/assets/img/custom/piece/code.svg',
+    }
+
+    private BRANCH_ITEM_DETAILS = {
+        type: ActionType.BRANCH,
+        name: 'Branch',
+        description: 'Decide what happens based on an if condition result',
+        logoUrl: '/assets/img/custom/piece/branch.svg',
+    }
+
+    private LOOP_ITEM_DETAILS = {
+        type: ActionType.LOOP_ON_ITEMS,
+        name: 'Loop',
+        description: 'Loop on a list of items',
+        logoUrl: '/assets/img/custom/piece/loop.svg',
+    }
+
+    private EMPTY_TRIGGER_ITEM_DETAILS = {
+        type: TriggerType.EMPTY,
+        name: 'Trigger',
+        description: 'Choose a trigger',
+        logoUrl: '/assets/img/custom/piece/empty-trigger.svg',
+    }
+
+    private CORE_ACTIONS_DETAILS: FlowItemDetails[] = [
+        this.BRANCH_ITEM_DETAILS,
+        this.LOOP_ITEM_DETAILS,
+        this.CODE_ITEM_DETAILS
+    ]
+    private clearCache$ = new Subject<void>();
+    private cachedPieceMetadata: { [key: string]: Observable<PieceMetadataModel> } = {};
+
+    constructor(private authenticationService: AuthenticationService, private http: HttpClient) { }
+
+
+    clearCache() {
+        this.clearCache$.next();
+    }
+
+    listPieces(request: ListPiecesRequestQuery): Observable<PieceMetadataModelSummary[]> {
+        return this.http.get<PieceMetadataModelSummary[]>(`${environment.apiUrl}/pieces`, {
+            params: {
+                includeHidden: request.includeHidden ? 'true' : 'false',
+                ...spreadIfDefined('searchQuery', request.searchQuery),
+                ...spreadIfDefined('suggestionType', !isNil(request.searchQuery) && request.searchQuery.length > 3 ? request.suggestionType : undefined)
+            }
+        });
+    }
+
+    listInstalledPieces(): Observable<PieceMetadataModelSummary[]> {
+        return this.listPieces({
+            includeHidden: false
+        }).pipe(map(pieces => pieces.filter(piece => !isNil(piece.projectId))))
+    }
+
+    private listCorePieces(searchQuery: string | undefined, suggestionType: SuggestionType): Observable<PieceMetadataModelSummary[]> {
```

Comment:
rename to listCoreCategoryPieces or make a generic function where you pass the category and return the pieces based upon it, otherwise a newcomer will thing this is some sort of special piece same for apps